File: android/src/main/kotlin/com/mytiki/sdk/capture/receipt/capacitor/rsp/RspCoupon.kt
/*
 * Copyright (c) TIKI Inc.
 * MIT license. See LICENSE file in the root directory.
 */

package com.mytiki.sdk.capture.receipt.capacitor.rsp

import com.getcapacitor.JSObject
import com.microblink.core.Coupon
import org.json.JSONObject

/**
 * Represents a parsed coupon extracted from a receipt during the RSP (Receipt Scanning Processor) process.
 *
 * @property coupon The [Coupon] object containing coupon information.
 */
class RspCoupon(coupon: Coupon) : Rsp {
    private val type: String?
    private val amount: RspFloatType?
    private val sku: RspStringType?
    private val description: RspStringType?
    private val relatedProductIndex: Int

    init {
        type = coupon.typeToString()
        amount = RspFloatType.opt(coupon.amount())
        sku = RspStringType.opt(coupon.sku())
        description = RspStringType.opt(coupon.description())
        relatedProductIndex = coupon.relatedProductIndex()
    }

    /**
     * Converts the [RspCoupon] object to a JSON representation.
     *
     * @return A [JSONObject] containing the coupon information.
     */
    override fun toJS(): JSObject =
        JSObject()
            .put("type", type)
            .put("amount", amount?.toJS() ?: JSObject())
            .put("sku", sku?.toJS())
            .put("description", description?.toJS())
            .put("relatedProductIndex", relatedProductIndex)

    companion object {
        /**
         * Creates an optional [RspCoupon] object from a [Coupon].
         *
         * @param coupon The [Coupon] object to convert.
         * @return An [RspCoupon] object if [coupon] is not null, otherwise null.
         */
        fun opt(coupon: Coupon?): RspCoupon? =
            if (coupon != null) RspCoupon(coupon) else null
    }
}

File: ios/Plugin/Rsp/RspCoupon.swift
/*
 * RspCoupon Struct
 * Copyright (c) TIKI Inc.
 * MIT license. See LICENSE file in root directory.
 */

import Foundation
import Capacitor
import BlinkReceipt
import BlinkEReceipt

/**
 Represents a response containing coupon information for the ReceiptCapture plugin.

 This struct is used to encapsulate coupon details, including its type, amount, SKU, description, and related product index, in response to plugin calls.
 */
struct RspCoupon: Rsp {
    /// The type of the coupon.
    private let type: UInt?
    
    /// The coupon amount, if available.
    private let amount: Float?
    
    /// The SKU (Stock Keeping Unit) associated with the coupon, if available.
    private let sku: String?
    
    /// The description or text associated with the coupon, if available.
    private let description: String?
    
    /// The index of the related product associated with this coupon.
    private let relatedProductIndex: Int

    /**
     Initializes an `RspCoupon` struct based on a `BRCoupon` object.

     - Parameter coupon: A `BRCoupon` object containing coupon information.
     */
    init(coupon: BRCoupon) {
        type = coupon.couponType.rawValue
        amount = coupon.couponAmount.value
        sku = coupon.couponSku.value
        description = coupon.description
        relatedProductIndex = coupon.relatedProductIndex
    }

    /**
     Converts the `RspCoupon` struct into a dictionary suitable for use in plugin response data.

     - Returns: A dictionary representing coupon details, including its type, amount, SKU, description, and related product index, in a format suitable for a Capacitor plugin call result.
     */
    func toPluginCallResultData() -> Capacitor.PluginCallResultData {
        var result = JSObject()
        result["type"] = type as JSValue?
        result["amount"] = amount
        result["sku"] = sku
        result["description"] = description
        result["relatedProductIndex"] = relatedProductIndex
        return result
    }
}
